deliver/lib/deliver/app_screenshot_validator.rb

```diff
@@ -0,0 +1,103 @@
+require 'fastimage'
+
+module Deliver
+  class AppScreenshotValidator
+    # A simple structure that holds error information as well as format error message consistently
+    # Set `true` to `to_skip` when just needing to skip uploading rather than causing a crash.
+    class ValidationError
+      # Constants that can be given to `type` param
+      INVALID_SCREEN_SIZE = 'Invalid screen size'.freeze
+      UNACCEPTABLE_DEVICE = 'Not an accepted App Store Connect device'.freeze
+      INVALID_FILE_EXTENSION = 'Invalid file extension'.freeze
+      FILE_EXTENSION_MISMATCH = 'File extension mistmaches its image format'.freeze
+
+      attr_reader :type, :path, :debug_info, :to_skip
+
+      def initialize(type: nil, path: nil, debug_info: nil, to_skip: false)
+        @type = type
+        @path = path
+        @debug_info = debug_info
+        @to_skip = to_skip
+      end
+
+      def to_s
+        "#{to_skip ? '🏃 Skipping' : '🚫 Error'}: #{path} - #{type} (#{debug_info})"
+      end
+
+      def inspect
+        "\"#{type}\""
+      end
+    end
+
+    # Access each array by symbol returned from FastImage.type
+    ALLOWED_SCREENSHOT_FILE_EXTENSION = { png: ['png', 'PNG'], jpeg: ['jpg', 'JPG', 'jpeg', 'JPEG'] }.freeze
+
+    # Validate a screenshot and inform an error message via `errors` parameters. `errors` is mutated
+    # to append the messages and each message should contain the corresponding path to let users know which file gets the error.
+    #
+    # @param screenshot [AppScreenshot]
+    # @param errors [Array<Deliver::AppScreenshotValidator::ValidationError>] Pass an array object to add validation errors when detecting errors
+    # @return [Boolean] true if given screenshot is valid
+    def self.validate(screenshot, errors)
+      # Given screenshot will be diagnosed and errors found are accumulated
+      errors_found = []
+
+      validate_screen_size(screenshot, errors_found)
+      validate_device_type(screenshot, errors_found)
+      validate_file_extension_and_format(screenshot, errors_found)
+
+      # Merge errors found into given errors array
+      errors_found.each { |error| errors.push(error) }
```

Comment:
To understand this code, I had to read how arguments are passed in ruby (as value, or as reference): https://robertheaton.com/2014/07/22/is-ruby-pass-by-reference-or-pass-by-value/ - just to realize that it's neither value or reference, but something like a mix of them 🤯 Very interesting! haha cool 😁

Reply:
Ah yes, it's correct. This is a bit tricky. Normally, this should be avoided as it's all about side-effect in a function. But API design-perspective, I wanted to return just a boolean rather than doing throw-catch dance. This allows me to accumulate errors found easily. I believe this is not that bad since this `validate` method is a simple class method. Using a small side effect for this purpose won't harm us that much.

Reply:
Updated document message to mention that the array passed will be mutated.